### ISSUE TITLE
Make it configurable whether "extra common precondition places" (may) exist in conflict tasks - and another aspect

### ISSUE THREAD

**Issue by jvoigtlaender** [fmidue/modelling-tasks#82]
See:
![image](https://user-images.githubusercontent.com/5853832/105699634-f719b900-5f07-11eb-9de7-a9b8f4de4dce.png)

So if t and t' are in conflict, there is at least one s such that both t and t' try to take tokens from s. There can also be another s' from which both t and t' try to take tokens. It could even be that said s' is also responsible for a/the conflict between t and t'. Per `uniqueConflictPlace` we can already control whether such an extra s' that is also a cause for the conflict can (or even, should) exist. But we have no control so far over whether s' may exist that is in the preconditions of both t and t', but which is not a cause for conflict between them (because s' has enough tokens to in principle let them both fire - except that they can't, because of s).

The proposal is to add such control.

For the `FindConflictTask`, that means that an extra option of type `Maybe Bool` would be set with the following semantics:
- `Nothing` - no extra constraints, behavior is exactly as up to now
- `Just False` - There must be no extra s'. That is, for the two transitions in conflict, all the places on which they both depend must also be cause for conflict.
- `Just True` - There must be an extra s'. That is, for the two transitions in conflict, there should be, beside the places on which they are conflicted, at least one additional place on which they both depend.

Of course, the semantics of `Just False` and `Just True` could also be exactly the other way round, depending on how the new option is named.

It should also be considered whether there is an interplay with the `uniqueConflictPlace` configuration option. Are all the possible combinations of `uniqueConflictPlace` and the new option reasonable? Or maybe one wants to have an extra s' only if the original s is unique? Or similar considerations.

It would also be possible in principle to make a quantitative statement (rather than just an existential statement) about s'. Not just saying "should exist" vs. "must not exist", but instead "how many of such extra s' should exist". Maybe overkill.

For the `PickConflictTask`, additional questions arise. Probably all the options coming out of the above for `FindConflictTask` should be equally applied to the one Petri net in the `PickConflictTask` that is indeed the one with a conflict. But what about the other Petri net, that does not have a conflict? Should the new option also have an effect on that other Petri net? For example controlling/ensuring that there are some transitions that have common preconditions (even if non-conflicted ones)?

In fact, for `PickConflictTask` one might (even independently of the above new proposed option) want to be able to ensure that at least there are some transitions that have a common precondition. Otherwise, the `PickConflictTask` might become almost trivial. (If one of the two shown Petri nets does not even have a place that has outgoing edges to two different transitions, then it is very obvious that this is the one without a conflict. So students can pick the other one as their answer, without even actually looking for concrete conflicts there.)

So overall, there might be more rethinking needed here, about the existing `uniqueConflictPlace` option and one or two additional option fields.

**Comment by jvoigtlaender** [comment]
A related but still somewhat different aspect:

The current task designs do not - I think - really (explicitly) test the students' understanding that two transitions can only be in conflict if each individually is activated. That is, if we have two transitions t1 and t2, and they have a common precondition place, and together they try to take more tokens out of that place than there are present, then these transitions t1 and t2 are nevertheless *not* in conflict if one of them happens to be prevented from firing (possibly for a completely unrelated reason, another place with too few tokens).

It could be worthwhile to be able to enforce that students have to cope with exactly such a situation. For example, having to find a conflict in a situation where there are two pairs of transitions where each pair has a common precondition place with too few tokens for them firing together, but only one of these pairs is really a conflict, because in the other pair at least one of the two transitions cannot even fire on its own.

**Comment by marcellussiegburg** [comment]
I came up with the following
```Alloy
fun commonPreconditions[t1, t2 : Transitions] : set Places {
  flow.Int.t1 & flow.Int.t2
}

pred showConflict[places : some Places,activatedTrans : set Transitions, transition1, transition2 : Transitions] {
  [...]
  #commonPreconditions[transition1, transition2] > #places    // 1. force additional common preconditions
  let ts = Transitions - transition1 - transition2 |          // 2. 
    some t1 : ts | one t2 : ts - t1 |                         // 2. some different transitions t1, t2
      let ps = commonPreconditions[t1, t2] {                  // 2.
        #ps > #places                                         // 2i. having more than #places common preconditions
        all p : ps | p.tokens >= p.flow[t1] and p.tokens >= p.flow[t2] // 2ii.
        some p : ps | p.tokens < plus[p.flow[t1], p.flow[t2]] // 2ii. having something that looks like a conflict
      }
```
This enables constraining a lot of these parts
1.  - remove line for old behaviour
     - as is: forces other common preconditions for the conflict
     - set to `= #places` to force no other common preconditions for the conflict
2.  - remove lines for old behaviour
     - set to `no t1, t2 : ts | let ps = commonPreconditions[t1, t2] | #ps > 1 and all p : ps | p.tokens >= p.flow[t1] and p.tokens >= p.flow[t2]` to disable such distractors
     - as is: forces another pair of transitions with multiple common preconditions and looking like the other conflict
     - refine another pair of transitions
     1. - as is: for it having more common preconditions as conflict places
         - set to `= #places` for it having as many common preconditions as conflicting places
     2. - as is: to make it look like it is a conflict
         - change lines to `all p : ps | p.tokens >= plus[p.flow[t1], p.flow[t2]]` to make it (look like it is) concurrent

This enables already most of the aforementioned features. But maybe there is discussion about some of these aspects required:
- In my opinion it does not interfere with `uniqueConflictPlace` directly, as it allows additional common preconditions for the conflicting transitions. I would call this configuration option `addConflictCommonPreconditions`.
- I would call the second property `addConflictDistractors` (as `Maybe Bool) with additional Properties
  - `addConflictDistractorPrecondition` (as `Bool`) for 2i. which may be set only if `addConflictDistractors` is set
  - `conflictDistractorAs` (as `Bool`) for 2ii. which may be set only if `addConflictDistractors` is set.
- One could argue, if 2. should be limited to exactly one pair of such Transitions t1, t2.
- One could argue, if 2i. should be `#ps > #commonPreconditions[transition1, transition2]`
- One could argue, if it should be enforceable if common preconditions of `t1` and(/or) `t2` should be distinct to `places` (or common preconditions of `transition1` and `transition2`).

This is a PetriNet obtained by using above settings:
<img width="810" alt="image" src="https://user-images.githubusercontent.com/531939/154302860-2a394097-aad7-422e-b10f-1e35fba54982.png">

*Edit: fix second remark on 2ii.*

**Comment by jvoigtlaender** [comment]
I cannot fully digest all the alternatives here right now (or probably anytime soon).

Maybe it's most productive that you go with what you have thought of, and we put it to the test of actual student solution attempts.

In particular if you can "guarantee" that the current/old behavior remains reproducible by just "not using the new options" in a certain sense.

Just from superficially reading the above (also the code, but without context of what exactly else would by around when these constraints are checked), I had a few thoughts/concerns (but maybe they would be obsolete with more thinking):
- Maybe `#commonPreconditions[transition1, transition2] > #places` would better be expressed as `commonPreconditions[transition1, transition2] not in places`? (Not sure, though, what this does to efficiency; whether Alloy is better at set comparison than at counting and integer comparison.)
- Oh, the other bits are really obsolete. I thought for some moments that `all p : ps | p.tokens >= p.flow[t1] and p.tokens >= p.flow[t2]` in this context would be equivalent to `activated[t1] and activated[t2]`. But no, and that is exactly the point, due to only looking at common precondition places. (I assume there is some context in the `[...]` part that ensures that really just the pair `transition1, transition2` is an actual conflict?)

**Comment by marcellussiegburg** [comment]
> Maybe `#commonPreconditions[transition1, transition2] > #places` would better be expressed as `commonPreconditions[transition1, transition2] not in places`?

Good catch. I defined common preconditions as every place both depend on (so including the `places` involved in the conflict; therefore your expression is not fitting in this case). But `some (commonPreconditions[transition1, transition2] - places)` is bringing speedup as well.

> I assume there is some context in the [...] part that ensures that really just the pair transition1, transition2 is an actual conflict?

Yes, there are constraints already for this. For the sake of the above code we can assume: The only conflict existing is made of `transition1`, `transition2` and `places`.

**Comment by jvoigtlaender** [comment]
> therefore your expression is not fitting in this case

I don't get that.

Isn't it the case that `places` is a subset of `commonPreconditions[transition1, transition2]`?

If so, then `commonPreconditions[transition1, transition2] not in places` means that `commonPreconditions[transition1, transition2]` is a strict super set of `places` (because `not in` means `\notsubseteq`). Which makes it semantically equivalent to `some (commonPreconditions[transition1, transition2] - places)`.

Arguably, the semantics of `not in` is a bit non-evident (just from reading the operator) in the case that the left set is non-singleton. Actually, one could also, equivalently, write `not (commonPreconditions[transition1, transition2] = places)` here (I think).

**Comment by marcellussiegburg** [comment]
> Isn't it the case that `places` is a subset of `commonPreconditions[transition1, transition2]`?

It is the case! I misinterpreted the symbol here. You are of course right. Nevertheless, I think, I will stick with the `some` version. In this case in fact all 3 variants could be used.

**Comment by jvoigtlaender** [comment]
There are now quite a few distractors implemented for conflict tasks. I don't have enough of an overview to see whether this is covered:
- Having (requiring) that one (or the) conflict-contributing place has more than one token on it.

**Comment by marcellussiegburg** [comment]
No this is not yet covered. Probably a new configuration Option `inducingPlaceHasMultipleTokens` is useful.
